### DISCUSSION TITLE
TLS hook

### DISCUSSION THREAD

**Discussion by vaibhavmurkute** [emqx/emqx#7433] [Q&A] [answered]
Hi,

Does any of the auth plugins for EMQX broker provide hooks on TLS handshake? i.e. the hook that would allow custom validation of client certificates during TLS handshake. If yes, could you help point me to that code logic.

Thanks.

**Comment by zmstone** [comment, marked as answer]
@vaibhavmurkute not directly as a part of the handshake.
You can opt in to use certificate common name as clientid or username, then you can apply a lot different checks on it using the currently supported auth plugins

**Reply by vaibhavmurkute** [reply]
@zmstone thank you for the quick reply.
Does EMQX use Erlang's TLS libraries for the handshake? If applicable, could you help point me to the EMQX code that calls Erlang's TLS hooks/libraries?
(Assuming Erlang provides hooks on TLS, e.g. certificate-verify)

**Reply by zmstone** [reply]
Hi @vaibhavmurkute 

Yes, EMQX uses Erlang's TLS lib.
on the current master branch
here is where essentially the ssl options are put together and passed down to the esockd lib which manages the ssl listeners.
https://github.com/emqx/emqx/blob/8f5812e61fba66c4c05939fd8cbc5d4056cf7ab2/apps/emqx/src/emqx_listeners.erl#L463-L471

**Reply by vaibhavmurkute** [reply]
Thank you @zmstone . This is helpful.
Does EMQX have plans for enabling custom verification of certificates during TLS handshake, anytime soon?
Would you be open to upstream PRs for this use case? (possibly using Erlang's [custom_verify()](https://www.erlang.org/doc/man/ssl.html#type-custom_verify) hook)

**Reply by zmstone** [reply]
We do not have it in our plan, but PRs are welcome.
